extensions/webstorage/lib.rs

```diff
@@ -51,82 +46,68 @@ pub fn get_declaration() -> PathBuf {
   PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_webstorage.d.ts")
 }
 
-struct WebStorageConnectionResource(Connection);
+struct LocalStorage(Connection);
+struct SessionStorage(Connection);
 
-impl Resource for WebStorageConnectionResource {
-  fn name(&self) -> Cow<str> {
-    "webStorage".into()
-  }
-}
-
-pub fn op_webstorage_open(
+fn get_webstorage(
   state: &mut OpState,
   persistent: bool,
-  _zero_copy: Option<ZeroCopyBuf>,
-) -> Result<u32, AnyError> {
-  let connection = if persistent {
-    let path = state.try_borrow::<LocationDataDir>().ok_or_else(|| {
-      DomExceptionNotSupportedError::new(
-        "LocalStorage is not supported in this context.",
-      )
-    })?;
-    std::fs::create_dir_all(&path.0)?;
-    Connection::open(path.0.join("local_storage"))?
+) -> Result<&Connection, AnyError> {
+  let conn = if persistent {
+    if state.try_borrow::<LocalStorage>().is_none() {
+      let path = state.try_borrow::<LocationDataDir>().ok_or_else(|| {
+        DomExceptionNotSupportedError::new(
+          "LocalStorage is not supported in this context.",
+        )
+      })?;
+      std::fs::create_dir_all(&path.0)?;
+      let conn = Connection::open(path.0.join("local_storage"))?;
+      state.put(LocalStorage(conn));
+    }
+
+    &state.borrow::<LocalStorage>().0
   } else {
-    Connection::open_in_memory()?
+    if state.try_borrow::<SessionStorage>().is_none() {
+      let conn = Connection::open_in_memory()?;
+      state.put(SessionStorage(conn));
+    }
+
+    &state.borrow::<SessionStorage>().0
   };
 
-  connection.execute(
+  conn.execute(
     "CREATE TABLE IF NOT EXISTS data (key VARCHAR UNIQUE, value VARCHAR)",
     params![],
   )?;
```

Comment:
This is now called in every op. Please move it to the initialization procedures.

Reply:
Done